### ISSUE TITLE
Northstar freezes when launched through Viper (1.9.5)

### ISSUE THREAD

**Issue by ASpoonPlaysGames** [0neGal/viper#150]
[This PR](https://github.com/R2Northstar/NorthstarLauncher/pull/237) that was merged for 1.9.5 (and reverted for 1.9.6) is causing issues in the game when launching through Viper.

From initial investigations, this seems to potentially be a problem with how Viper creates the NorthstarLauncher process?

(I am going to create a similar issue in NorthstarLauncher, just figured I'd put it here as well)

**Comment by Alystrasz** [comment]
It was also noted on Discord that the issue might be related to Viper launching Northstar with no dedicated console window.

**Comment by 0neGal** [comment]
May or may not be caused by `run()` (`child_process.spawn`) and switching to `exec()` (`child_process.exec`) may fix the issue?

I'm not entirely certain, I don't currently have a working Wine prefix with Titanfall installed, so I can't test this yet, but I'll have a look at getting that fixed and then debugging this...

That is if this is even a problem through Wine, who knows, once I'm able to get a working prefix and further a PR I'd need some confirmation on whether it made any difference on Windows

Besides this, the Linux launch support PR also has some problems with launching the game and switching to `exec` may fix it there as well. I've just not had the time to look into it recently :p

**Comment by 0neGal** [comment]
> It was also noted on Discord that the issue might be related to Viper launching Northstar with no dedicated console window.
> We can try to run Northstar independently of Viper: https://nodejs.org/api/child_process.html#child_processspawncommand-args-options 

I was making my other reply before this came in, ig the mobile client doesn't live update replies...

`.spawn()` is what we're currently using afaik...

I do want to point out I haven't tested the Windows build beyond launching it though Wine to check for launch errors, so I wouldn't actually know about the console not appearing...

**Comment by Alystrasz** [comment]
I was pointing out the `detached` option.
I'll take a look at it!

**Comment by 0neGal** [comment]
> I was pointing out the `detached` option.
> I'll take a look at it!

Ah, didn't realize, try that one first then, and if its still broken, then see if `.exec()` changes anything...

**Comment by GeckoEidechse** [comment]
Even if the console stays attached, I'm confused why it causes an issue in the first place...

**Comment by GeckoEidechse** [comment]
Actually, @ASpoonPlaysGames, your PR addresses two things at once. Coloured logs and [removing `freopen`](https://github.com/R2Northstar/NorthstarLauncher/pull/237/files#diff-e44bbfecf4ccfd08fe5c840400ca0e34d7aee1fafa6371fa3393831391819d01R237-R241). Maybe it's actually related to the second part?

That seems more plausible IMO ^^

**Comment by ASpoonPlaysGames** [comment]
> Actually, @ASpoonPlaysGames, your PR addresses two things at once. Coloured logs and [removing `freopen`](https://github.com/R2Northstar/NorthstarLauncher/pull/237/files#diff-e44bbfecf4ccfd08fe5c840400ca0e34d7aee1fafa6371fa3393831391819d01R237-R241). Maybe it's actually related to the second part?
> 
> That seems more plausible IMO ^^

Potentially, yes. However using freopen isn't something I want to do, since it is the cause of the logging issues we used to have where the output wouldn't appear in the console

**Comment by 0neGal** [comment]
If a build could be made with just one of them, and then test it, that'd be nice...

But it does sound implausible that colors in an arbitrary window would affect Viper's ability to launch the game...

**Comment by pg9182** [comment]
> However using freopen isn't something I want to do, since it is the cause of the logging issues we used to have where the output wouldn't appear in the console

It's also part of why nswrap is necessary for the docker image (though the workaround isn't terribly complicated)

**Comment by pg9182** [comment]
For Windows, see https://docs.microsoft.com/en-us/windows/console/creation-of-a-console.

**Comment by Alystrasz** [comment]
> Ah, didn't realize, try that one first then, and if its still broken, then see if `.exec()` changes anything...

I went the `exec` way.
Test builds are [available here](https://github.com/Alystrasz/viper/releases/tag/freezes-fix), can someone try to launch `v1.9.5` with one of them? (it worked on my computer, but now I'm sure of nothing)

**Comment by ASpoonPlaysGames** [comment]
I can confirm that on that build, my game doesn't freeze anymore, however on my normal viper installation, the game still does freeze

**Comment by Alystrasz** [comment]
> I can confirm that on that build, my game doesn't freeze anymore, however on my normal viper installation, the game still does freeze

Yeah, I still gotta do a PR to merge the fix, and @0neGal must create a new release to deploy said fix to all clients.
PR incoming :)

**Comment by 0neGal** [comment]
Just to verify, even though I doubt this wouldn't be the case, but everything works in both `1.9.5` and `1.9.6` with the move to `exec()` right?

**Comment by Alystrasz** [comment]
> Just to verify, even though I doubt this wouldn't be the case, but everything works in both `1.9.5` and `1.9.6` with the move to `exec()` right?

After playing with `v1.9.5` files, I forced Viper to update my game to `v1.9.6`, and got to play with seemingly no problem.

**Comment by 0neGal** [comment]
Sounds good, I'll have a look at actually merging tomorrow, along with making a release, I'm not gonna try and publish a release hours before going to bed :p

**Comment by ASpoonPlaysGames** [comment]
> Just to verify, even though I doubt this wouldn't be the case, but everything works in both `1.9.5` and `1.9.6` with the move to `exec()` right?

It would appear so. I'm not testing on 1.9.6 technically, I'm just building NorthstarLauncher with and without the PR that broke things, things seem to work fine for me when using `exec()`
